conjure_oxide/src/solvers/kissat.rs

```diff
@@ -0,0 +1,526 @@
+use crate::ast::Domain::BoolDomain;
+use crate::ast::Expression::{And, Not, Or, Reference};
+use crate::ast::{DecisionVariable, Domain, Expression, Model, Name};
+use std::collections::HashMap;
+use String;
+
+type Error = String;
+
+struct CNF {
+    pub clauses: Vec<Vec<i32>>,
+    variables: HashMap<Name, i32>,
+    next_ind: i32,
+}
+
+impl CNF {
+    pub fn new() -> CNF {
+        CNF {
+            clauses: Vec::new(),
+            variables: HashMap::new(),
+            next_ind: 1,
+        }
+    }
+
+    pub fn get_variables(&self) -> Vec<&Name> {
+        let mut ans: Vec<&Name> = Vec::new();
+
+        for key in self.variables.keys() {
+            ans.push(key);
+        }
+
+        return ans;
+    }
+
+    pub fn get_index(&self, var: &Name) -> Option<i32> {
+        return match self.variables.get(var) {
+            None => None,
+            Some(idx) => Some(*idx),
+        };
+    }
+
+    pub fn get_name(&self, ind: i32) -> Option<&Name> {
+        for key in self.variables.keys() {
+            let idx = self.get_index(key)?;
+            if idx == ind {
+                return Some(key);
+            }
+        }
+        return None;
+    }
+
+    pub fn add_variable(&mut self, var: &Name) {
+        self.variables.insert(var.clone(), self.next_ind);
+        self.next_ind += 1;
+    }
+
+    pub fn has_variable<T: HasVariable>(&self, value: T) -> bool {
+        value.has_variable(self)
+    }
+
+    pub fn add_clause(&mut self, vec: &Vec<i32>) -> Result<(), Error> {
+        for idx in vec {
+            if !self.has_variable(idx.abs()) {
+                return Err(format!("Variable with index {idx} not found!"));
+            }
+        }
+        self.clauses.push(vec.clone());
+        return Ok(());
+    }
+
+    pub fn add_expression(&mut self, expr: &Expression) -> Result<(), Error> {
+        for row in self.handle_expression(expr)? {
+            self.add_clause(&row)?;
+        }
+        return Ok(());
+    }
+
+    pub fn as_expression(&self) -> Result<Expression, Error> {
+        let mut expr_clauses: Vec<Expression> = Vec::new();
+
+        for clause in &self.clauses {
+            expr_clauses.push(self.clause_to_expression(clause)?);
+        }
+
+        // ToDo (gs248) We should probably flatten the result first. x -> And( [ Or([x]) ] ) is a bit silly.
```

Comment:
I think having a nice standard result is fine, some other code can do improvement / cleanup if it wants (there is lots of ways to clean up SAT!)